app/controllers/cases_controller.rb

```diff
@@ -278,37 +285,173 @@ def destroy
     )
     rus.upload!
+
+    @s3_direct_post = S3Uploader.s3_direct_post_for_case(@case, 'responses')
+
+    case rus.result
+    when :blank
+      flash.now[:alert] = t('alerts.response_upload_blank?')
+      render :upload_responses
+    when :error
+      flash.now[:alert] = t('alerts.response_upload_error')
+      render :upload_responses
+    when :ok
+      flash[:notice] = t('notices.response_uploaded')
+      set_permitted_events
+      redirect_to case_path @case
+    end
+  end
+
+  def upload_response_and_approve
+    authorize @case
+
+    @s3_direct_post = S3Uploader.s3_direct_post_for_case(@case, 'responses')
+    @approval_action = 'approve'
+    @case = @case.decorate
+  end
+
+  #rubocop:disable Metrics/MethodLength
+  def execute_upload_response_and_approve
+    authorize @case, :upload_response_and_approve?
+
+    service_params = {
+      kase: @case,
+      current_user: current_user,
+      action: 'upload-approve',
+      uploaded_files: params[:uploaded_files],
+      upload_comment: params[:upload_comment],
+      is_compliant: true,
+      bypass_message: nil,
+      bypass_further_approval: false
+    }
+    if params.key?(:bypass_approval)
+      service_params.merge!(
+        bypass_message: params[:bypass_approval][:bypass_message],
+        bypass_further_approval: !params[:bypass_approval][:press_office_approval_required] == 'true'
+      )
+    end
+
+    service = ResponseUploaderService.new(service_params)
+    service.upload!
+
+    @case = @case.decorate
+    case service.result
+    when :blank
+      @s3_direct_post = S3Uploader.s3_direct_post_for_case(@case, 'responses')
+      flash.now[:alert] = t('alerts.response_upload_blank?')
+      render :upload_response_and_approve
+    when :error
+      @s3_direct_post = S3Uploader.s3_direct_post_for_case(@case, 'responses')
+      flash.now[:alert] = t('alerts.response_upload_error')
+      render :upload_response_and_approve
+    when :ok
+      flash[:notice] = t('notices.response_uploaded')
+      set_permitted_events
+      redirect_to case_path @case
+    end
+  end
+  #rubocop:enable Metrics/MethodLength
+
+  def upload_response_and_return_for_redraft
+    authorize @case
     @s3_direct_post = S3Uploader.s3_direct_post_for_case(@case, 'responses')
+    @approval_action = 'approve'
+    @case = @case.decorate
+  end
+
+  def upload_response_and_return_for_redraft_action
+    authorize @case, :upload_response_and_return_for_redraft?
+
+    rus = ResponseUploaderService.new(
+      kase: @case,
+      current_user: current_user,
+      action: 'upload-redraft',
+      uploaded_files: params[:uploaded_files],
+      upload_comment: params[:upload_comment],
+      is_compliant: params[:draft_compliant] == 'yes',
+      bypass_message: nil,
+      bypass_further_approval: false
+    )
+    rus.upload!
+
+    @case = @case.decorate
     case rus.result
     when :blank
+      @s3_direct_post = S3Uploader.s3_direct_post_for_case(@case, 'responses')
```

Comment:
There's a few uses of the strings `'approve'` and `'redraft'` etc. I feel a bit nervous about them as the codebase grows and more developers join. Should be Constants/Enums or something a bit more robust IMO. Too big a job for this PR?

Reply:
Out of scope for this PR but something to consider.